core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go

```diff
@@ -18,107 +18,96 @@ import (
-	// CleanUpInterval is the interval for cleaning up block maps
-	CleanUpInterval = 15 * time.Minute
-	// ChannelSize represents the channel size for head broadcaster
-	ChannelSize = 20
+	// cleanUpInterval is the interval for cleaning up block maps
+	cleanUpInterval = 15 * time.Minute
+	// channelSize represents the channel size for head broadcaster
+	channelSize = 20
 )
 
-type BlockKey struct {
+type blockKey struct {
 	block int64
-	hash  common.Hash
+	hash  string
 }
 
-func (bk *BlockKey) getBlockKey() ocr2keepers.BlockKey {
+func (bk *blockKey) getBlockKey() ocr2keepers.BlockKey {
 	return ocr2keepers.BlockKey{
 		Number: ocr2keepers.BlockNumber(bk.block),
-		Hash:   bk.hash,
+		Hash:   common.HexToHash(bk.hash),
 	}
 }
 
 type BlockSubscriber struct {
-	sync                  utils.StartStopOnce
-	mu                    sync.RWMutex
-	ctx                   context.Context
-	cancel                context.CancelFunc
-	hb                    httypes.HeadBroadcaster
-	lp                    logpoller.LogPoller
-	headC                 chan BlockKey
-	unsubscribe           func()
-	subscribers           map[int]chan ocr2keepers.BlockHistory
-	blocksFromPoller      map[int64]common.Hash
-	blocksFromBroadcaster map[int64]common.Hash
-	maxSubId              int
-	lastClearedBlock      int64
-	lastSentBlock         int64
-	blockHistorySize      int64
-	lggr                  logger.Logger
+	sync             utils.StartStopOnce
+	mu               sync.RWMutex
+	ctx              context.Context
+	cancel           context.CancelFunc
+	hb               httypes.HeadBroadcaster
+	lp               logpoller.LogPoller
+	headC            chan *evmtypes.Head
+	unsubscribe      func()
+	subscribers      map[int]chan ocr2keepers.BlockHistory
+	blocks           map[int64]string
+	maxSubId         int
+	lastClearedBlock int64
+	lastSentBlock    int64
+	blockHistorySize int64
+	lggr             logger.Logger
 }
 
 func NewBlockSubscriber(hb httypes.HeadBroadcaster, lp logpoller.LogPoller, blockHistorySize int64, lggr logger.Logger) *BlockSubscriber {
 	return &BlockSubscriber{
-		hb:                    hb,
-		lp:                    lp,
-		headC:                 make(chan BlockKey, ChannelSize),
-		subscribers:           map[int]chan ocr2keepers.BlockHistory{},
-		blocksFromPoller:      map[int64]common.Hash{},
-		blocksFromBroadcaster: map[int64]common.Hash{},
-		blockHistorySize:      blockHistorySize,
-		lggr:                  lggr.Named("BlockSubscriber"),
+		hb:               hb,
+		lp:               lp,
+		headC:            make(chan *evmtypes.Head, channelSize),
+		subscribers:      map[int]chan ocr2keepers.BlockHistory{},
+		blocks:           map[int64]string{},
+		blockHistorySize: blockHistorySize,
+		lggr:             lggr.Named("BlockSubscriber"),
 	}
 }
 
-func (hw *BlockSubscriber) getBlockRange(ctx context.Context) ([]uint64, error) {
-	h, err := hw.lp.LatestBlock(pg.WithParentCtx(ctx))
+func (hw *BlockSubscriber) getBlockRange() ([]uint64, error) {
```

Comment:
not sure i follow the motivation for removing ctx argument from here

Reply:
since hw struct already has a ctx so i am just using it directly

Reply:
Using a context like that is a bit hacky, let's use the one we get as a parameter (although we will pass the same context probably)

Reply:
done